File: TraningAPI/View/PrefecturesListView/PrefecturesListViewController.m
//
//  PrefecturesListViewController.m
//  TraningAPI
//
//  Created by 高橋慧 on 2021/06/23.
//  Copyright © 2021 k-takahashi. All rights reserved.
//

#import "PrefecturesListViewController.h"

@interface PrefecturesListViewController ()

@end

@implementation PrefecturesListViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.presenter = [[PrefecturesPresenter alloc] initWithViewControllerDelegate:self];
    [self.presenter getPrefecturesModel:true];
    self.prefecturesListTableView.delegate = self;
    self.prefecturesListTableView.dataSource = self;
    [self.prefecturesListTableView registerNib:[UINib nibWithNibName:PrefecturesListTableViewCell.cellIdentifer bundle:nil] forCellReuseIdentifier:PrefecturesListTableViewCell.cellIdentifer];
    self.reloadButton = [[UIBarButtonItem alloc] initWithImage:[UIImage systemImageNamed:@"arrow.clockwise"] style:UIBarButtonItemStyleDone target:self action:@selector(reloadTableView)];

    self.navigationItem.rightBarButtonItem = self.reloadButton;
    self.regionModel = [[RegionModel alloc] init];
}
- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:true];
    [self.delegate viewWillAppear];
}
- (void)reloadTableView {
    [self.presenter getPrefecturesModel:false];
}
- (void)alert:(NSError *)error {
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"エラーが発生しました" message:error.localizedDescription preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:^(UIAlertAction *action){
        [self.presenter getPrefecturesModel:false];
    }];
    [alert addAction:okAction];
    [self presentViewController:alert animated:true completion:nil];
}

- (void)setList {
    [self.prefecturesListTableView reloadData];
}

- (void)setLastUpdate:(NSString *)lastUpdate {
    self.lastUpdateLabel.text = lastUpdate;
    [self.reloadButton setEnabled:true];
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    NSNumber *indexRow = self.regionModel.prefecturesID[indexPath.section][indexPath.row];
    DetailsViewController *to = [StoryBoard performDetailsView:self prefecturesID:indexRow.integerValue + 1];
    [StoryBoard push:self toViewController:to];
    [self.delegate viewWillAppear];
}
- (UIView *)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section {
    UITableViewHeaderFooterView *header = [[UITableViewHeaderFooterView alloc] init];
    UITapGestureRecognizer *gesture = [[UITapGestureRecognizer alloc] initWithTarget: self action: @selector(touchHeader:)];
    [header addGestureRecognizer:gesture];
    header.tag = section;
    return header;
}
/// アコーディオン表示制御
- (void)touchHeader:(UITapGestureRecognizer *)sender {
    NSIndexSet *index = [[NSIndexSet alloc] initWithIndex:sender.view.tag];
    NSMutableArray *nameIsHidden = [self.regionModel.nameIsHidden mutableCopy];
    if (self.regionModel.nameIsHidden[sender.view.tag].boolValue) {
        nameIsHidden[sender.view.tag] = [NSNumber numberWithBool: false];
    } else {
        nameIsHidden[sender.view.tag] = [NSNumber numberWithBool: true];
    }
    self.regionModel.nameIsHidden = [nameIsHidden copy];
    [self.prefecturesListTableView beginUpdates];
    [self.prefecturesListTableView reloadSections:index withRowAnimation:UITableViewRowAnimationAutomatic];
    [self.prefecturesListTableView endUpdates];
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    NSString *name = self.regionModel.region[section];
    return name;
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return self.regionModel.region.count;
}

- (NSInteger)tableView:(nonnull UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    if (self.regionModel.nameIsHidden[section].boolValue) {
        return 0;
    } else {
        NSArray *prefectures = self.regionModel.prefecturesID[section];
        return prefectures.count;
    }
}

- (nonnull UITableViewCell *)tableView:(nonnull UITableView *)tableView cellForRowAtIndexPath:(nonnull NSIndexPath *)indexPath {
    PrefecturesListTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:PrefecturesListTableViewCell.cellIdentifer forIndexPath:indexPath];
    NSNumber *indexRow = self.regionModel.prefecturesID[indexPath.section][indexPath.row];
    [cell nameLabelConfigure:[self.presenter getPrefectureName:indexRow.integerValue]];
    [cell infectedConfigure:[self.presenter getCasesNumber:indexRow.integerValue]];
    return cell;
}



@end

File: TraningAPI/View/PrefecturesListView/PrefecturesListViewController.swift
//
//  ViewController.swift
//  TraningAPI
//
//  Created by 高橋慧 on 2021/04/14.
//  Copyright © 2021 k-takahashi. All rights reserved.
//

import UIKit

class PrefecturesListViewController: UIViewController{
    
    private var presenter: PrefecturesPresenterInput!
    weak var delegate: RootContainerDelegate?
    
    @IBOutlet weak var prefecturesListTableView: UITableView!{
        didSet{
            prefecturesListTableView.delegate = self
            prefecturesListTableView.dataSource = self
            prefecturesListTableView.register(UINib(nibName: PrefecturesListTableViewCell.cellIdentifer, bundle: nil), forCellReuseIdentifier: PrefecturesListTableViewCell.cellIdentifer)
        }
    }
    @IBOutlet weak var lastUpdateLabel: UILabel!
    @IBOutlet weak var reloadButton: UIButton!{
        didSet{
            reloadButton.addTarget(self,action: #selector(reloadTableView), for: .touchUpInside)
            reloadButton.isEnabled = false
        }
    }
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        presenter = PrefecturesPresenter(delegate: self)
        print("Presenterに情報取得の指示")
        presenter.getPrefecturesModel(updateComp: true)
    }
    deinit {
        print("prefectureslist", #function)
    }
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(true)
        print(#function)
        self.navigationController?.navigationBar.isHidden = true
        self.delegate?.viewWillAppear()
        
        print(self.parent?.parent)
    }
    @objc func reloadTableView(){
        presenter.getPrefecturesModel(updateComp: false)
    }
}
extension PrefecturesListViewController:PrefecturesView{
    func setList() {
        print("Presenterから再描画の指示があったのでVIewを更新する")
        prefecturesListTableView.reloadData()
    }
    func setlastUpdate(lastUpdate:String){
        lastUpdateLabel.text = lastUpdate
        reloadButton.isEnabled = true
    }
    func alert(error: APIError){
        let alert = UIAlertController(title: "エラーが発生しました", message: error.massegeDescription, preferredStyle: .alert)
        let okAction: UIAlertAction = UIAlertAction(title: "OK", style: .default, handler:{
            (action: UIAlertAction) -> Void in
            self.presenter.getPrefecturesModel(updateComp: false)
        })
        alert.addAction(okAction)
        self.present(alert, animated: true, completion: nil)
    }
}
extension PrefecturesListViewController:UITableViewDelegate{
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        //遷移とPresenterから情報を渡す
        /*
         Viewではあくまでも依頼を流すだけなら、Presenterに書くべき処理？
         */
        if let indexRow = RegionModel.Region(rawValue: indexPath.section)?.prefecturesID[indexPath.row]{
            StoryBoard.perform(id: indexRow+1, to: Board.details, from: self)
            self.delegate?.viewWillAppear()
        }
        
    }
    func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
        let header = UITableViewHeaderFooterView()
        let gesture = UITapGestureRecognizer(target: self,
                                             action: #selector(touchHeader))
        header.addGestureRecognizer(gesture)
        header.tag = section
        return header
    }
    @objc func touchHeader(_ sender: UITapGestureRecognizer) {
        guard let section = sender.view?.tag else { return }
        RegionModel.nameIsHidden[section] = !RegionModel.nameIsHidden[section]
        prefecturesListTableView.beginUpdates()
        prefecturesListTableView.reloadSections([section], with: .automatic)
        prefecturesListTableView.endUpdates()
    }
}
extension PrefecturesListViewController: UITableViewDataSource{
    
    func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        guard let region = RegionModel.Region(rawValue: section) else { return "" }
        return region.name
    }
    
    func numberOfSections(in tableView: UITableView) -> Int {
        return RegionModel.regions.count
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        guard let region = RegionModel.Region(rawValue: section) else { return 0 }
       
        if RegionModel.nameIsHidden[section] {
            return 0
        }else{
            return region.prefecturesID.count
        }
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        guard let cell = tableView.dequeueReusableCell(withIdentifier: PrefecturesListTableViewCell.cellIdentifer, for: indexPath) as? PrefecturesListTableViewCell else{
            return UITableViewCell()
        }
        if let indexRow = RegionModel.Region(rawValue: indexPath.section)?.prefecturesID[indexPath.row]{
            cell.nameLabelConfigure(name: presenter.getPrefectureName(index: indexRow ))
            cell.infectedConfigure(number: presenter.getCasesNumber(index: indexRow))
        }
        return cell
    }
    
}
